modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java

```diff
@@ -0,0 +1,449 @@
+                if (currVal != null)
+                    return true;
+
+                if (!cursor.next())
+                    return false;
+
+                currVal = cursor.get();
+
+                return true;
+            }
+
+            /** {@inheritDoc} */
+            @Override protected IgniteBiTuple<K, V> onNext() {
+                if (currVal == null)
+                    if (!hasNext())
+                        throw new NoSuchElementException();
+
+                IndexRow row = currVal;
+
+                currVal = null;
+
+                K k = (K) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().key(), keepBinary, false);
+                V v = (V) CacheObjectUtils.unwrapBinaryIfNeeded(coctx, row.cacheDataRow().value(), keepBinary, false);
+
+                return new IgniteBiTuple<>(k, v);
+            }
+        };
+    }
+
+    /** Get index to run query by specified description. */
+    private Index index(GridCacheContext cctx, IndexQueryDesc idxQryDesc) throws IgniteCheckedException {
+        Class<?> valCls = idxQryDesc.valCls() != null ? loadValClass(cctx, idxQryDesc.valCls()) : null;
+
+        String tableName = cctx.kernalContext().query().tableName(cctx.name(), valCls);
+
+        if (tableName == null)
+            return null;
+
+        // Find index by specified name.
+        if (idxQryDesc.idxName() != null) {
+            String name = idxQryDesc.idxName();
+
+            if (!QueryUtils.PRIMARY_KEY_INDEX.equals(name))
+                name = name.toUpperCase();
+
+            String schema = cctx.kernalContext().query().schemaName(cctx);
+
+            IndexName idxName = new IndexName(cctx.name(), schema, tableName, name);
+
+            Index idx = idxProc.index(idxName);
+
+            if (idx == null)
+                return null;
+
+            return checkIndex(idxProc.indexDefinition(idx.id()), idxQryDesc.criteria()) ? idx : null;
+        }
+
+        // Try get index by list of fields to query.
+        // Check all indexes by cache to find best index match: count of index fields equals to count of index criteria fields.
+        Collection<Index> idxs = idxProc.indexes(cctx);
+
+        for (Index i: idxs) {
+            IndexDefinition idxDef = idxProc.indexDefinition(i.id());
+
+            if (!tableName.equals(idxDef.idxName().tableName()))
+                continue;
+
+            if (checkIndex(idxDef, idxQryDesc.criteria()))
+                return i;
+        }
+
+        return null;
+    }
+
+    /** Checks that specified index matches index query criteria. */
+    private boolean checkIndex(IndexDefinition idxDef, List<IndexQueryCriterion> criteria) {
+        if (criteria.size() > idxDef.indexKeyDefinitions().size())
+            return false;
+
+        for (int i = 0; i < criteria.size(); i++) {
+            if (!idxDef.indexKeyDefinitions().get(i).name().equalsIgnoreCase(criteria.get(i).field()))
```

Comment:
Order of fields matters, as it's possible to create index (A, B) and (B, A). Also in case we have index (A, B) user can specify query for (B) only. It's illegal. User has to start with most significant field. So I think we should be restrictive there.